src/coreclr/jit/valuenum.cpp

```diff
@@ -9760,6 +9927,64 @@ bool Compiler::fgValueNumberHelperCall(GenTreeCall* call)
         {
             VNFunc vnf = fgValueNumberJitHelperMethodVNFunc(helpFunc);
 
+            // We have a special case: a helper call that represents a cast.
+            // For VN purposes, we do not care that it is actually a call, so
+            // value number it like any normal cast.
+            // TODO-Cleanup: consider moving the expansion to casting helpers
+            // from morph to lower to avoid having to do this special handling
```

Comment:
I tried to prototype this, but ran into problems:
1) Removing this logic from morph and adding it to "simple lowering" means that things that call morph in the backend can no longer trust the returned trees to be "correct for sequencing and lowering".
2) Moving this to `LowerNode` is problematic because we cannot easily introduce calls there as the "simple lowering" phase, which computes the outgoing args has already run.

Overall, I am not very happy with the special-casing of helpers here & in morph, but not doing it would make the VN changes mostly redundant for FP<->integer casts.